codec-redis/src/main/java/io/netty/handler/codec/redis/RedisEncoder.java

```diff
@@ -0,0 +1,154 @@
+    }
+
+    private static void writeIntegerMessage(ByteBufAllocator allocator, IntegerRedisMessage msg, List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        out.add(numberToByteBuf(msg.value()));
+        out.add(EOL.slice());
+    }
+
+    private static void writeBulkStringHeader(ByteBufAllocator allocator, BulkStringHeaderRedisMessage msg,
+                                              List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+        } else {
+            out.add(numberToByteBuf(msg.bulkStringLength()));
+            out.add(EOL.slice());
+        }
+    }
+
+    private static void writeBulkStringContent(BulkStringRedisContent msg, List<Object> out) {
+        out.add(msg.content().retain());
+        if (msg instanceof LastBulkStringRedisContent) {
+            out.add(EOL.slice());
+        }
+    }
+
+    private static void writeFullBulkStringMessage(ByteBufAllocator allocator, FullBulkStringRedisMessage msg,
+                                               List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+        } else {
+            ByteBuf content = msg.content();
+            out.add(numberToByteBuf(content.readableBytes()));
+            out.add(EOL.slice());
+            out.add(content.retain());
+        }
+        out.add(EOL.slice());
+    }
+
+    /**
+     * Write array header only without body. Use this if you want to write arrays as streaming.
+     */
+    private static void writeArrayHeader(ByteBufAllocator allocator, ArrayHeaderRedisMessage msg, List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+        } else {
+            out.add(numberToByteBuf(msg.length()));
+        }
+        out.add(EOL.slice());
+    }
+
+    /**
+     * Write full constructed array message.
+     */
+    private static void writeArrayMessage(ByteBufAllocator allocator, ArrayRedisMessage msg, List<Object> out) {
+        out.add(allocator.ioBuffer(TYPE_LENGTH).writeByte(msg.type().value()));
+        if (msg.isNull()) {
+            out.add(NULL_BULK_STRING.slice());
+            out.add(EOL.slice());
+        } else {
+            out.add(numberToByteBuf(msg.children().size()));
+            out.add(EOL.slice());
+            for (RedisMessage child : msg.children()) {
+                writeRedisMessage(allocator, child, out);
+            }
+        }
+    }
+
+    private static ByteBuf numberToByteBuf(long value) {
+        if (value == -1L) {
+            return NULL_BULK_STRING.slice();
+        }
+        return Unpooled.wrappedBuffer(Long.toString(value).getBytes(CharsetUtil.US_ASCII));
```

Comment:
It sounds good. Then, I think we can provide the (cache strategy or factory) interface and a fixed default implementation instead of current global static RedisMessageFactory. And, we can use it as a constructor parameter of encoder/decoder and remove global states. (Current factory has global states for cache.)